holmes/plugins/toolsets/grafana/trace_parser.py

```diff
@@ -0,0 +1,222 @@
+        """Calculate duration in milliseconds"""
+        return (
+            self.end_time - self.start_time
+        ) / 1_000_000  # Convert nanoseconds to milliseconds
+
+
+def decode_id(encoded_id: str) -> str:
+    """Decode base64 IDs to a hex string for easier reading"""
+    return base64.b64decode(encoded_id).hex()
+
+
+def build_span_hierarchy(trace_data: Dict) -> List[Span]:
+    # Step 1: Extract all spans and create span objects
+    all_spans = {}
+
+    for batch in trace_data["batches"]:
+        # Extract service name and other resource attributes
+        service_name = "unknown"
+        resource_attributes = {}
+
+        for attr in batch["resource"]["attributes"]:
+            key = attr["key"]
+            value = list(attr["value"].values())[0]
+            resource_attributes[key] = value
+
+            if key == "service.name":
+                service_name = value
+
+        for scope_spans in batch["scopeSpans"]:
+            for span_data in scope_spans["spans"]:
+                span_id = decode_id(span_data["spanId"])
+
+                # Get parent span ID if it exists
+                parent_span_id = None
+                if "parentSpanId" in span_data:
+                    parent_span_id = decode_id(span_data["parentSpanId"])
+
+                # Create a Span object
+                span = Span(
+                    span_id=span_id,
+                    parent_span_id=parent_span_id,
+                    name=span_data["name"],
+                    service_name=service_name,
+                    start_time=int(span_data["startTimeUnixNano"]),
+                    end_time=int(span_data["endTimeUnixNano"]),
+                    attributes={
+                        attr["key"]: list(attr["value"].values())[0]
+                        for attr in span_data.get("attributes", [])
+                    },
+                    events=span_data.get("events", []),
+                    resource_attributes=resource_attributes,
+                )
+
+                all_spans[span_id] = span
+
+    # Step 2: Build the hierarchy by connecting parents and children
+    root_spans = []
+
+    for span in all_spans.values():
+        if span.parent_span_id is None:
+            # This is a root span
+            root_spans.append(span)
+        else:
+            # Add this span as a child to its parent
+            if span.parent_span_id in all_spans:
+                all_spans[span.parent_span_id].children.append(span)
+
+    return root_spans
+
+
+def get_trace_timestamps(root_spans):
+    """Calculate the start and end times for the entire trace"""
+    if not root_spans:
+        return None, None
+
+    min_start = min(span.start_time for span in root_spans)
+    max_end = max(span.end_time for span in root_spans)
+
+    # Check all children recursively
+    for root_span in root_spans:
+        for child in root_span.children:
+            child_min, child_max = get_span_timestamps_recursive(child)
```

Comment:
Suggestion: Change `get_trace_timestamps` to `to get_spans_timestamps` and make the recurstion on child.spans. Thay way you will need only one implementation of this function and can reuse logic.

Reply:
Simplified the file :+1: